### PR TITLE
Explicit async collections instead of "threads"

### PR DESCRIPTION
**Pull request by xbrianh**
This allows finer-grained control of the number of async operations used in client applications.

### CODE REVIEW

File: gs_chunked_io/reader.py [hunk 1]
```diff
@@ -97,45 +95,37 @@ def write(self, *args, **kwargs):
...
-        for chunk_number in range(reader.number_of_chunks):
-            for chunk in chunks.consume_finished():
-                yield chunk
-                break
-            chunks.put(fetch_chunk, chunk_number)
-        for chunk in chunks.consume():
+    for chunk_number in range(reader.number_of_chunks):
+        for chunk in async_set.consume_finished():
             yield chunk
+            break
```
Comment (DailyDreaming) [inline gs_chunked_io/reader.py:130]: It's not immediately obvious to me why there's a break here.  Could a comment be added?

File: gs_chunked_io/writer.py [hunk 2]
```diff
@@ -43,13 +43,11 @@ def __init__(self,
         self._buffer = bytearray()
         self._current_part_number = 0
         self._closed = False
-        if threads is not None:
-            assert 1 <= threads
-            max_workers = max(threads, 8)
-            self.executor: Optional[ThreadPoolExecutor] = ThreadPoolExecutor(max_workers=max_workers)
-            self.future_chunk_uploads = AsyncSet(self.executor, concurrency=threads)
+        self.future_chunk_uploads: Optional[AsyncSet] = None
```
Comment (DailyDreaming) [inline gs_chunked_io/writer.py]: I think that this:

```
self.future_chunk_uploads: Optional[AsyncSet] = None
if async_set is not None:
    self.future_chunk_uploads = async_set
else:
    self.future_chunk_uploads = None
```

may be the same as:

```
self.future_chunk_uploads: Optional[AsyncSet] = async_set
```

**Review by DailyDreaming — APPROVED** [review]
LGTM.  See comments.

File: gs_chunked_io/writer.py (hunk 2 above)
Comment (xbrianh) [reply gs_chunked_io/writer.py]: Thanks! will fix.

File: gs_chunked_io/reader.py (hunk 1 above)
Comment (xbrianh) [reply gs_chunked_io/reader.py:128]: will fix

Comment (xbrianh) [reply gs_chunked_io/reader.py:130]: Added comment:
```
# Breaking after the first yield allows us to add more downloads to the pot without           
# waiting for the client to complete potentially time-consuming operations.
```

**Pull request merged**